How To Build Globally Distributed Applications with Azure Cosmos DB (#1777)
pulumi/docs: content/blog/how-to-build-globally-distributed-applications-with-azure-cosmos-db-and-pulumi/index.md

---
title: "How To Build Globally Distributed Applications with Azure Cosmos DB and Pulumi"
authors: ["mikhail-shilkov"]
tags: ["Azure"]
meta_image: "meta.png"
meta_desc: "A reusable component to build highly-available, low-latency applications on Azure"
date: "2019-09-24"
---

> In my [previous article](https://www.pulumi.com/blog/globally-distributed-serverless-application-in-100-lines-of-code-infrastructure-included/), I shared how easy it is to create a globally distributed, highly-available, low-latency application with Azure Functions, Azure Cosmos DB, and Pulumi.

Reviewer: Nit: Mind using `relref` for links within the site?

```suggestion
In my [previous article]({{< relref "/blog/globally-distributed-serverless-application-in-100-lines-of-code-infrastructure-included" >}}), I shared how easy it is to create a globally distributed, highly-available, low-latency application with Azure Functions, Azure Cosmos DB, and Pulumi.
```

> In my [previous article](https://www.pulumi.com/blog/globally-distributed-serverless-application-in-100-lines-of-code-infrastructure-included/), I shared how easy it is to create a globally distributed, highly-available, low-latency application with Azure Functions, Azure Cosmos DB, and Pulumi.

Reviewer 2: ```suggestion
In a [previous blog post](https://www.pulumi.com/blog/globally-distributed-serverless-application-in-100-lines-of-code-infrastructure-included/), I shared how easy it is to create a globally distributed, highly-available, low-latency application with Azure Functions, Azure Cosmos DB, and Pulumi.
```

Today, I want to show how the same approach can be generalized for any cloud compute service, including containers, virtual machines, and serverless functions.

<!--more-->

## E-commerce Example

Let’s consider an example of an e-commerce website which targets a worldwide audience. The architecture of such a website should address several challenges.

> ### Multiple teams

Reviewer 2: ```suggestion
### Multiple Teams
```

A large e-commerce website is a sophisticated software application and consists of many subcomponents. Many engineers in multiple teams work on it simultaneously.

A typical approach is to organize engineers into smaller teams and make each team autonomous and responsible for one or several business domains. This strategy leads to microservices architecture when the application is built out of loosely coupled components. Each component represents a "vertical slice": a team owns functionality across the vertical stack, from its user interface and service API to the data storage layer and underlying infrastructure.

Here is an example of such a breakdown:

![Multiple components on a single web page](website.png)

{{% figcaption %}}Multiple components on a single web page{{% /figcaption %}}

> ### Diversity of technology options and choices

Reviewer 2: ```suggestion
### Diversity of Technology Decisions
```

One of the guiding principles of microservices architecture is that each team makes decisions regarding which tools and technologies to use. They are free to choose programming languages, databases, or cloud services that are the best fit for the service. In practice, some restrictions do apply---for instance, the teams might need to share the same cloud provider.

In our hypothetical example, the whole product is hosted on Azure cloud, however other aspects differ:

- Product Page team runs JavaScript serverless functions and stores data in a document database
- Shopping Cart team uses C# and .NET Core, packages the code in a Docker container, and uses event sourcing to store every action ever made by users while shopping
- Pricing Engine team has a mix of Java and C++ code with a bunch of third-party libraries and deploys the services to Ubuntu VMs that use MongoDB drivers for storage

> ### Global presence

Reviewer 2: ```suggestion
### Global Presence
```

Regardless of the tech, the user-facing services must serve customers across the globe and serve them fast. This requirement demands a geo-distributed infrastructure so that user requests could be handled by the nearest data center to minimize the latency and ensure excellent user experience.

Each team has to solve the following problems in a way which fits their tech stack:

- Distribute the data across the globe
- Run their compute workload in each region next to the data
- Manage all this infrastructure in a reliable way

Let's consider these goals and possible solutions.

## Azure Cosmos DB

> Distributing the data across many data centers while keeping them consistent and available is one of the hardest problems in computer science and engineering. Delegating this challenge to a specialized service or product is usually a sensible idea.

Reviewer 2: ```suggestion
Distributing data across many data centers while keeping them consistent and available is one of the hardest problems in computer science and engineering. Delegating this challenge to a specialized service or product is usually a sensible idea.
```

![Global application with Virtual Machines](vmscalesets.png)

{{% figcaption %}}Global application with Virtual Machines{{% /figcaption %}}

Each team has the same global structure of infrastructure, including a Traffic Manager front-end and a Cosmos DB data store. At the same time, the per-region compute infrastructure differs substantially.

Instead of copying the common pieces of infrastructure between the teams, we can think of a reusable abstraction which could be shared by every service.

![Global application pattern](globalapp.png)

{{% figcaption %}}Global application pattern{{% /figcaption %}}

For this blog post, I created a custom Pulumi component called CosmosApp which implements such abstraction and simplifies the provisioning of global web applications. The component creates distributed Cosmos DB resources, as well as the front-end routing component, while allowing pluggable compute layer implementation.

## Cosmos App with Azure Functions

Let's look at an example of how the component can be used. Creating a globally-distributed application requires three steps:

> ### List the regions

Reviewer 2: ```suggestion
### List the Regions
```

We define a configuration value to contain a comma-separated list of Azure regions that we’ll deploy the application to.

```
$ pulumi config set locations WestUS,WestEurope,SouthEastAsia,SouthBrazil,AustraliaCentral
```

This value can be loaded and parsed into a JavaScript array:

```ts
const locations = new pulumi.Config().require("locations").split(",");
```

> ### Build a region template

Reviewer 2: ```suggestion
### Build a Region Template
```

We define a factory function to create the infrastructure in each location. Its signature is a bit unusual in that we return a function from another function. I'll explain the motivation in the next example. For now, it's important to understand the following:

> - We get a Cosmos account and a location as input arguments.

Reviewer 2: ```suggestion
- We get a Cosmos account and a location as input arguments
```

> - We create a Function App in the specified location with connection settings pointing to the local Cosmos DB replica.

Reviewer 2: ```suggestion
- We create a Function App in the specified location with connection settings pointing to the local Cosmos DB replica
```

> - We return the identifier of the Function App so that a Traffic Manager endpoint could point to it.

Reviewer 2: ```suggestion
- We return the identifier of the Function App so that a Traffic Manager endpoint could point to it
```

The Shopping Cart team will have to follow the same steps but define their compute infrastructure accordingly.

A significant difference to the Functions-based example is the fact that the Container-based application needs some infrastructure to be shared across regions. The top block of the factory function defines an instance of Azure Container Registry, builds a Docker image, and puts that image into the Registry:

```ts
function buildShoppingCartApp({ cosmosAccount, database, container }: GlobalContext) {

    const registry = new azure.containerservice.Registry("global", {
        resourceGroupName: resourceGroup.name,
        adminEnabled: true,
        sku: "Premium",
    }, opts);

    const dockerImage = new docker.Image("node-app", {
        imageName: pulumi.interpolate`${registry.loginServer}/mynodeapp:v1.0.0`,
        build: {
            context: "./container",
        },
        registry: {
            server: registry.loginServer,
            username: registry.adminUsername,
            password: registry.adminPassword,
        },
    });

    return ({ location }: RegionalContext) => {
        // ... see the next sample
    };
}
```

> Now, the factory function that builds per-region infrastructure can directly reference the Registry and the image to deploy an Azure Container Instances group. Note how registry and dockerImage are used as any other TypeScript variable inside the closure factory function:

Reviewer 3: ```suggestion
Now, the factory function that builds per-region infrastructure can directly reference the Registry and the image to deploy an Azure Container Instances group. Note how `registry` and `dockerImage` are used as any other TypeScript variable inside the closure factory function:
```

                        port: 80,
                        protocol: "TCP",
                    }],
                    environmentVariables: {
                        ENDPOINT: cosmosAccount.endpoint,
                        MASTER_KEY: cosmosAccount.primaryMasterKey,
                        DATABASE: database.name,
                        COLLECTION: container.name,
                        LOCATION: location,
                    },
                },
            ],
            ipAddressType: "public",
            dnsNameLabel: `acishop-${location}`,
        });

        return {
            url: group.fqdn,
        };
    };
}
```

Finally, a `CosmosApp` is defined:

```ts
const aci = new CosmosApp("aci", {
    resourceGroup,
    locations,
    databaseName: "cartdb",
    containerName: "items",
    factory: buildShoppingCartApp,
    enableMultiMaster: true,
});
```

Shopping Cart workload expects both reads and writes from the end-users, so the team decides to enable multi-master support in their Cosmos DB instance.

[ [Full Example](https://github.com/pulumi/examples/tree/master/azure-ts-cosmosapp-component/aci.ts) ]

## Cosmos App with Virtual Machine Scale Sets

> The Pricing Engine team has to go through the same steps but employ IaaS-based infrastructure to run application code. The factory function creates a dozen of resources related to virtual machines, load balancing, networking, and auto-scaling. You may define as many resources as needed in the code, as long as it returns a pointer to the proper resource to link the endpoint.

Reviewer 3: ```suggestion
The Pricing Engine team has to go through the same steps but employ IaaS-based infrastructure to run application code. The factory function creates a dozen resources related to virtual machines, load balancing, networking, and auto-scaling. You may define as many resources as needed in the code, as long as it returns a pointer to the proper resource to link the endpoint.
```

Author: My dictionary suggests "of" here, as some examples that I see online. Is there a rule about that?

Reviewer 3: It's generally either "dozens of donuts" or "a dozen donuts", but not "a dozen of donuts." It's fine, however, to say, "A bag of donuts".  

Honestly, I have no idea why, and I'd love to have a better explanation than that. Maybe it has something to do with "dozen" being more of a counting word, like "twelve" (so like "twelve donuts" rather than "twelve of donuts". But then, "twelve of _those_ donuts" is okay?  

Ugh. English 🙄 .
